test/integration/deployment/deployment_test.go

```diff
@@ -1379,3 +1380,118 @@ func TestDeploymentScaleSubresource(t *testing.T) {
 	// Use the scale subresource to scale the deployment down to 0
 	testScalingUsingScaleSubresource(t, tester, 0)
 }
+
+func TestReplicaSetOrphaningAndAdoptionWhenLabelsChange(t *testing.T) {
```

Comment:
We don't actually care about RS label changes. What we do care about is the ownership between Deployment and its RSes. Some RS labels changes trigger OwnerRef changes implicitly.

Reply:
The latter is what this test is for: it should be possible to "steal" an RS by changing its labels to no longer match the Deployment selector. This test verifies that the Deployment automatically orphans the RS in this case.

Reply:
This test is added to ensure the change on `sync.go` is working as expected. There is an integration test named `TestGeneralReplicaSetAdoption` that tests against ownership between Deployment and its RSes.

When writing the deployment basic integration tests, I found out a bug that the deployment does not orphan an RS when the RS's `.Labels` field is changed. A workaround is to change both selector or template labels of the RS alongside with the `.Labels` field change. However, this is not expected.

It was later on found out that collision avoidance mechanism does not trigger when the user makes a change that does not modify the RS's PodTemplateSpec (e.g., `.Labels` field change). This is because the mechanism _only_ assumes that if a RS's PodTemplateSpec is semantically deep equal with the deployment's PodTemplateSpec, then the RS is the new RS for the deployment.

The issue that I linked with this PR is confusing - it still narrates about `.Labels` change without generalizing into the case of intact PodTemplateSpec. I will update the issue accordingly, and ping you there when it is ready. For this PR, I have updated the comments in `sync.go`. To clarify, I will add a comment about what the `TestReplicaSetOrphaningAndAdoptionWhenLabelsChange` test is testing against.